### DISCUSSION TITLE
Filtering features with WebGL when Draw the lines 🙏

### DISCUSSION THREAD

**Discussion by romprokop** [openlayers/openlayers#15380] [Ideas]
Is it possible to expand the functionality of drawing the lines using WebGL:
- filtering features like for points; (https://openlayers.org/en/latest/examples/filter-points-webgl.html)
- add stroke-line-join-size (especially for 'round')
- own event for 'stroke-line-join' symbol

To draw a path for 10k+ coordinates, and display points on the lines corner and arrow, I create a Layer Group with 3 independent layers, which consumes significant memory. I could achieve the same result using vertex shaders, but it isn't easy, and examples are missed in the documentation

**Comment by jahow** [comment]
Hi! Thanks for your interest in the library.

> filtering features like for points; (https://openlayers.org/en/latest/examples/filter-points-webgl.html)

This should already be doable if you specify a `filter` on the style used for the lines. Granted, there's no example showing that.

> add stroke-line-join-size (especially for 'round')

What would that do exactly? A round join usually has a diameter equal to the line width. Would you have a visual example?

> own event for 'stroke-line-join' symbol

Same, could you provide a visual example? If the idea is to place symbols at each vertex of a line string, I think this would be very interesting but probably not achieved with such a property. IMO this would typically be a geometry transformation, i.e. turn a line string into a collection of points at its vertices.

**Reply by romprokop** [reply]
@jahow Thanks for replying🙏 Without a doubt, OpenLayers is an awesome library!!

I made the path rendered without WebGL, this is a layer group with lines, dots, arrows, and start/end flags. I already implemented the same using WebGl but it still consumes a significant memory. If the path contain a 1-10k signals its ok, but for more than 50k and 10-30 similar path, but GPU almost full and uses 10-16 GB from swap

<img width="1373" alt="Screenshot 2023-11-28 at 13 48 07" src="https://github.com/openlayers/openlayers/assets/36002116/03a16800-0c32-48eb-9c03-318a3c63878c">

**Reply by jahow** [reply]
High memory usage of the WebGL renderer is a known issue and definitely on my list of things to address before the end of the year.

As for showing symbols on segments and vertices, for now in WebGL you'd have no choice but to preprocess the geometries before adding them to the source.

Thanks for your kind words :slightly_smiling_face:

**Reply by romprokop** [reply]
Got it! I wish you inspiration in realizing your planned goals🤗
